File: Core kernel Yaiwes/OpenAgentd/code/web/src/api/terminal.ts
/**
 * Terminal WebSocket client.
 *
 * Connect flow:
 *   1. POST /api/terminal/ticket (desktop token / access
 *      key attached automatically by installDesktopAuth's fetch patch).
 *   2. Open ws(s)://…/api/terminal/ws?ticket=…&_token=… — the single-use
 *      ticket authorizes the session; `_token` additionally satisfies the
 *      middleware's defence-in-depth WS check when a token is configured.
 *
 * Wire protocol (JSON text frames):
 *   send: {type:'input', data} | {type:'resize', rows, cols}
 *   recv: {type:'output', data} | {type:'exit'}
 */

import { apiUrl } from './base-url'
import { withTokenParam } from './auth'

export interface TerminalSocketCallbacks {
  onOutput: (data: string) => void
  onExit?: () => void
  onError?: (err: Error) => void
  onClose?: () => void
}

export interface TerminalSocket {
  sendInput: (data: string) => void
  sendResize: (rows: number, cols: number) => void
  close: () => void
}

interface TicketResponse {
  ticket: string
  expires_in: number
}

/**
 * The PTY starts at the validated absolute project workspace path.
 */
export type TerminalTarget = { workspace: string }

export async function fetchTerminalTicket(
  target: TerminalTarget,
  rows = 24,
  cols = 80,
): Promise<string> {
  const source = { workspace: target.workspace }
  const res = await fetch(apiUrl('/terminal/ticket'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ ...source, rows, cols }),
  })
  if (!res.ok) {
    const detail = await res
      .json()
      .then((b: { detail?: string }) => b.detail)
      .catch(() => undefined)
    throw new Error(detail ?? `Terminal ticket failed (${res.status})`)
  }
  const body = (await res.json()) as TicketResponse
  return body.ticket
}

/** Derive the ws(s):// URL for a given /api path from the active base URL. */
export function terminalWsUrl(ticket: string): string {
  const httpUrl = new URL(
    withTokenParam(`${apiUrl('/terminal/ws')}?ticket=${encodeURIComponent(ticket)}`),
    window.location.origin,
  )
  httpUrl.protocol = httpUrl.protocol === 'https:' ? 'wss:' : 'ws:'
  return httpUrl.toString()
}

export async function connectTerminal(
  target: TerminalTarget,
  callbacks: TerminalSocketCallbacks,
  size?: { rows: number; cols: number },
): Promise<TerminalSocket> {
  const ticket = await fetchTerminalTicket(
    target,
    size?.rows ?? 24,
    size?.cols ?? 80,
  )
  const ws = new WebSocket(terminalWsUrl(ticket))

  ws.onmessage = (event: MessageEvent<string>) => {
    try {
      const msg = JSON.parse(event.data) as { type: string; data?: string }
      if (msg.type === 'output' && typeof msg.data === 'string') {
        callbacks.onOutput(msg.data)
      } else if (msg.type === 'exit') {
        callbacks.onExit?.()
      }
    } catch {
      callbacks.onError?.(new Error('Malformed terminal frame'))
    }
  }
  ws.onerror = () => callbacks.onError?.(new Error('Terminal connection error'))
  ws.onclose = () => callbacks.onClose?.()

  await new Promise<void>((resolve, reject) => {
    ws.onopen = () => resolve()
    const prevOnClose = ws.onclose
    ws.onclose = (ev) => {
      reject(new Error('Terminal connection refused'))
      prevOnClose?.call(ws, ev)
    }
  })
  // Restore steady-state close handler after successful open.
  ws.onclose = () => callbacks.onClose?.()

  return {
    sendInput: (data: string) => {
      if (ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({ type: 'input', data }))
      }
    },
    sendResize: (rows: number, cols: number) => {
      if (ws.readyState === WebSocket.OPEN) {
        ws.send(JSON.stringify({ type: 'resize', rows, cols }))
      }
    },
    close: () => ws.close(),
  }
}

File: Core kernel Yaiwes/OpenAgentd/code/app/api/routes/terminal.py
"""``/api/terminal`` — interactive PTY terminal over WebSocket.

Auth model (two layers):

1. **HTTP ticket issuance** — ``POST /api/terminal/ticket`` sits behind
   the normal HTTP auth (``DesktopTokenMiddleware``: desktop token or
   LAN access key when configured). It validates the workspace and
   returns a short-lived, single-use ticket.
2. **WS connect** — ``GET /api/terminal/ws?ticket=…`` upgrades to a
   WebSocket. Browsers can't attach ``Authorization`` headers to WS
   handshakes, so the ticket (never the long-lived token) rides the
   URL. Tickets expire in seconds and burn on first use, so a leaked
   URL from a log is worthless. ``DesktopTokenMiddleware`` additionally
   enforces the ``?_token=`` rule on all ``/api/*`` WS upgrades as
   defence in depth; the frontend appends both.

Wire protocol (JSON text frames):

    client → server: {"type": "input",  "data": "<utf-8 keystrokes>"}
                     {"type": "resize", "rows": N, "cols": N}
    server → client: {"type": "output", "data": "<utf-8 chunk>"}
                     {"type": "exit"}

The PTY runs **on the backend host** — for external/LAN servers that is
the remote machine, not the client. UI copy must make this clear.
"""

from __future__ import annotations

import asyncio
import codecs
import hmac
import secrets
import time

from fastapi import APIRouter, HTTPException, WebSocket, WebSocketDisconnect
from loguru import logger
from pydantic import BaseModel, Field

from app.services import agent_manager, terminal_service

router = APIRouter()


def _new_decoder() -> codecs.IncrementalDecoder:
    """A fresh stream-aware UTF-8 decoder for one WS connection's output.

    ``bytes.decode("utf-8", errors="replace")`` on each PTY read chunk
    independently is wrong: a multibyte character (any non-ASCII glyph —
    box-drawing, emoji, p10k's powerline separators) can straddle two
    ``os.read()`` calls, and decoding each half in isolation corrupts it
    into U+FFFD even though the concatenated bytes are perfectly valid.
    ``codecs.getincrementaldecoder`` buffers a trailing partial sequence
    and completes it on the next ``decode()`` call instead.
    """
    return codecs.getincrementaldecoder("utf-8")(errors="replace")


# ── Ticket store ─────────────────────────────────────────────────────────────

#: ticket → (validated_workspace, expiry_monotonic, rows, cols). Single-use;
#: popped on first WS connect attempt. Module-level like other in-process
#: registries (mirrors ``terminal_service._SESSIONS``). rows/cols ride the
#: ticket so the PTY is spawned at the client's actual terminal size
#: instead of a 24x80 default that visibly reflows the instant the first
#: client resize frame arrives.
_TICKETS: dict[str, tuple[str, float, int, int]] = {}

TICKET_TTL_SECONDS = 30.0
_MAX_PENDING_TICKETS = 32


class TicketRequest(BaseModel):
    """Exactly one cwd source: a coding workspace path OR a chat session id.

    - ``workspace`` (coding mode): client-supplied absolute path, validated
      through ``agent_manager.validate_workspace`` (must exist, blocklisted
      system roots rejected).
    - ``session_id`` (coding workspace): UUID only — the server derives the
      path from ``OPENAGENTD_WORKSPACE_DIR``, so no client-controlled path
      ever reaches the filesystem. Created lazily (the sandbox dir may not
      exist until an agent first writes to it).
    """

    workspace: str = Field(..., min_length=1)
    rows: int = Field(default=24, ge=1, le=1000)
    cols: int = Field(default=80, ge=1, le=4000)


class TicketResponse(BaseModel):
    ticket: str
    expires_in: float


def _prune_expired_tickets() -> None:
    now = time.monotonic()
    for ticket, (_ws, expiry, _rows, _cols) in list(_TICKETS.items()):
        if now > expiry:
            del _TICKETS[ticket]


@router.post("/ticket", response_model=TicketResponse)
async def issue_ticket(body: TicketRequest) -> TicketResponse:
    """Resolve the cwd source and mint a single-use WS connect ticket."""
    try:
        resolved = agent_manager.validate_workspace(body.workspace)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    _prune_expired_tickets()
    if len(_TICKETS) >= _MAX_PENDING_TICKETS:
        raise HTTPException(status_code=429, detail="Too many pending tickets.")

    ticket = secrets.token_urlsafe(32)
    _TICKETS[ticket] = (
        resolved,
        time.monotonic() + TICKET_TTL_SECONDS,
        body.rows,
        body.cols,
    )
    return TicketResponse(ticket=ticket, expires_in=TICKET_TTL_SECONDS)


class _RedeemedTicket(BaseModel):
    workspace: str
    rows: int
    cols: int


def _redeem_ticket(ticket: str) -> _RedeemedTicket | None:
    """Pop and return the workspace/size for *ticket* if valid, else None.

    Constant-time lookup: compare against all stored tickets so a
    timing probe cannot distinguish "unknown" from "known" tickets.
    """
    _prune_expired_tickets()
    matched: str | None = None
    for stored in _TICKETS:
        if hmac.compare_digest(stored, ticket):
            matched = stored
    if matched is None:
        return None
    workspace, expiry, rows, cols = _TICKETS.pop(matched)
    if time.monotonic() > expiry:
        return None
    return _RedeemedTicket(workspace=workspace, rows=rows, cols=cols)


# ── WebSocket ────────────────────────────────────────────────────────────────


@router.websocket("/ws")
async def terminal_ws(websocket: WebSocket) -> None:
    ticket = websocket.query_params.get("ticket", "")
    redeemed = _redeem_ticket(ticket) if ticket else None
    if redeemed is None:
        # Deny before accept → handshake fails with 403.
        await websocket.close(code=4401)
        return

    await websocket.accept()

    try:
        session = await terminal_service.create_session(
            workspace=redeemed.workspace, rows=redeemed.rows, cols=redeemed.cols
        )
    except (RuntimeError, OSError) as exc:
        # RuntimeError: MAX_SESSIONS cap (expected, user-actionable).
        # OSError: real spawn failure — e.g. the host's PTY table is
        # exhausted (``pty.openpty()`` raises "out of pty devices") or
        # exec of the shell binary failed. Both are terminal-session-only
        # failures, not ASGI/server bugs — report and close the socket
        # rather than letting the exception propagate and crash the
        # connection handler.
        logger.warning(
            "terminal_spawn_failed workspace={} err={}", redeemed.workspace, exc
        )
        await websocket.send_json({"type": "output", "data": f"\r\n{exc}\r\n"})
        await websocket.close(code=4429)
        return

    logger.info(
        "terminal_ws_connected session_id={} workspace={}",
        session.session_id,
        redeemed.workspace,
    )

    async def pty_to_ws() -> None:
        decoder = _new_decoder()
        while True:
            chunk = await session.read()
            if chunk is None:
                # Flush any partial sequence left buffered in the decoder
                # (a truncated multibyte tail at EOF) before signalling exit.
                tail = decoder.decode(b"", final=True)
                if tail:
                    await websocket.send_json({"type": "output", "data": tail})
                await websocket.send_json({"type": "exit"})
                return
            await websocket.send_json({"type": "output", "data": decoder.decode(chunk)})

    async def ws_to_pty() -> None:
        while True:
            msg = await websocket.receive_json()
            match msg.get("type"):
                case "input":
                    data = msg.get("data", "")
                    if isinstance(data, str) and data:
                        await session.write(data.encode("utf-8"))
                case "resize":
                    rows, cols = msg.get("rows"), msg.get("cols")
                    if isinstance(rows, int) and isinstance(cols, int):
                        session.resize(rows=rows, cols=cols)

    to_ws = asyncio.create_task(pty_to_ws())
    to_pty = asyncio.create_task(ws_to_pty())
    try:
        done, pending = await asyncio.wait(
            {to_ws, to_pty}, return_when=asyncio.FIRST_COMPLETED
        )
        for task in pending:
            task.cancel()
        for task in done:
            exc = task.exception()
            if exc is not None and not isinstance(
                exc, (WebSocketDisconnect, asyncio.CancelledError)
            ):
                logger.warning(
                    "terminal_ws_error session_id={} err={}", session.session_id, exc
                )
    except WebSocketDisconnect:
        pass
    finally:
        to_ws.cancel()
        to_pty.cancel()
        await session.close()
        try:
            await websocket.close()
        except Exception as exc:
            # Socket may already be closed by the client — nothing to do.
            logger.debug("terminal_ws_close_failed error={!r}", exc)
